### DISCUSSION TITLE
Appropriate Architecture for Generating Links?

### DISCUSSION THREAD

**Discussion by CalmLogarithm** [stellargraph/stellargraph#1809] [General] [answered]
I had a previous question about link prediction. Thanks again to those who took the time to respond. I'm not 100% sure that I'm going down the right path with link prediction so I wanted to describe my application and see if it's the right match.

I have a knowledge base. Right now I am training node embeddings. Eventually I'll have a system that produces embeddings for new, previously unseen nodes. What I'm trying to find is an architecture that can generate one or more links when provided a node embedding as input.

So which NN architecture(s), or problems, would be the most appropriate for me to study?

**Comment by huonw** [comment, marked as answer]
This sounds like link prediction. One common framing of link prediction is to predict a score for a link between a given pair of nodes existing. This is often done by computing embedding vectors for the two nodes, and then applying a classifier on top of the pair of vectors (this can also be done end-to-end, rather than computing embedding vectors and separately training). For instance, the cosine similarity of the two embeddings could be used as a (non-trainable) classifier.

"Generating" links could then be done by scoring the new node against other nodes, and taking the highest scoring links or all links with score above some threshold, or similar.  Notes:

- In the general setting, the scoring has to be done in a pairwise manner, meaning your chosen (unseen) node would have to be scored against each other node, one-by-one. In a graph with many nodes, this may be expensive, and so one might randomly sample nodes first, or have some approximate/intelligent scheme for choosing likely other nodes. 
- Another option is to choose a restricted classifier that allows for more efficient searches. For instance, by using cosine or L2 distance between embedding vectors, all the embedding vectors could be placed into an approximate nearest-neighbour library like https://github.com/nmslib/nmslib and then generating links for a node consists of computing the embedding vector and then doing a look-up for the nearest neighbours (i.e. the nodes with most similar embeddings)

There's several demos for link prediction: https://stellargraph.readthedocs.io/en/stable/demos/link-prediction/index.html . Ones that may be of particular relevance to your problem are the following, because they allow for predicting on unseen nodes:

- https://stellargraph.readthedocs.io/en/stable/demos/link-prediction/graphsage-link-prediction.html (this also happens to use the "ip" method for scoring edges, which is "inner-product", aka cosine distance)
- https://stellargraph.readthedocs.io/en/stable/demos/link-prediction/attri2vec-link-prediction.html

Does that help?

**Reply by CalmLogarithm** [reply]
That does help, thank you. I anticipated that scaling a pariwise comparison might be a problem.

After re-reading the GraphSAGE tutorial, I think I found what I was missing. It wasn't clear to me where the link embedding based on the node embeddings was coming from. I didn't see it defined. Obviously, it's the `stellargraph.layer.link_classification` method.

Here's an added layer of complexity: Link prediction is a boolean result, correct? What if I had attributes associated with my links that I wanted generated as well? For example, in my case I might have `FROG -[ IS-A ]-> ANIMAL` where FROG and ANIMAL are nodes and IS-A is a label that commonly occurs on links.

**Reply by huonw** [reply]
Yeah, the common framing is a single score from 0 (predicted link does not exist) to 1 (predicted link exists). It sounds like you're wanting to also do some attribute inference too, inferring which of the labels to apply (I'll assume there's `L` labels). I think there's two ways to do this:

1. Do binary link existence prediction first, and then do multi-class classification for the `L` different classes, for each of the predicted links.
2. Do multi-class classification with an additional "DOES NOT EXIST" label. That is, the output would be a softmax vector of size `L + 1`, and the first entry might be the score of the link not existing, and the rest of the entries represent the score of the link existing & having a certain label.

I don't have a good sense for which of these would work best, but I suspect 2 is more "end-to-end", and so may work better.

How does that sound?

**Reply by CalmLogarithm** [reply]
Both approaches sound promising. I agree that #2 wraps both components up in a single model nicely. For the input, would I want link embeddings?

**Reply by huonw** [reply]
Yeah, sort of: many of StellarGraph's models support automatically creating link embeddings from node embeddings and combining them in some way and then running those through a classifier, with the whole thing trainable end-to-end, rather than having to compute link embeddings and then train a classifier to run over them.

**Reply by CalmLogarithm** [reply]
That makes sense, thanks. Re-reading the GraphSAGE and HinSAGE link prediction examples helped me understand how it trains node embeddings and link embeddings together. Is there a way to extract the trainked link embeddings? Or, would I need to reconstruct them using the same method that I specified in the edge_embedding_method parameter?

**Reply by huonw** [reply]
I think it depends what you consider your link embeddings to be, and what the embedding method is. For instance, the `ip`/`dot` methods directly create an output of size 1 from the two node embedding vectors, and so there's no link embedding vector.

For the other methods, you could:

- either reproduce them from the node embedding vectors as you say
- pull the `stellargraph.layer.LinkEmbedding` layer out of the Keras `Model`'s `.layers` property and create a second `Model` with that as the output (but the same inputs). There's an example doing something somewhat similar (training a model via link prediction, but then creating a second model for extracting the _node_ embeddings) that might serve as inspiration: https://stellargraph.readthedocs.io/en/stable/demos/embeddings/graphsage-unsupervised-sampler-embeddings.html#Extracting-node-embeddings (the `inputs` param could just be `x_inp` and the `outputs`  param would be the `LinkEmbedding` layer mentioned above.)

**Reply by CalmLogarithm** [reply]
I just wanted to confirm that I was able to get this working with GraphSAGE and link_inference. Thanks again!
